File: TripKitUI/views/TKUITripSegmentsView.m
//
//  TKUITripSegmentsView.m
//  TripKit
//
//  Created by Adrian Schoenig on 20/01/2014.
//
//

#import "TKUITripSegmentsView.h"

#ifdef TK_NO_MODULE
#import "TripKit.h"
#import "TripKit/TripKit-Swift.h"
#else
@import TripKit;
#import "TripKitUI/TripKitUI-Swift.h"
#endif


#import "TKUIStyledLabel.h"
#import "UIColor+Variations.h"
#import "TKStyleManager+TripKitUI.h"
#import "UIView+Helpers.h"

#define SEGMENT_ITEM_ALPHA_SELECTED   1
#define SEGMENT_ITEM_ALPHA_DESELECTED 0.25

@implementation NSString (Sizing)

- (CGSize)sizeWithFont:(TKFont *)font maximumWidth:(CGFloat)maxWidth
{
  NSDictionary *atts = @{NSFontAttributeName: font};
  NSStringDrawingContext *context = [[NSStringDrawingContext alloc] init];
  context.minimumScaleFactor = 1;
  CGRect box = [self boundingRectWithSize:CGSizeMake(maxWidth, 0)
                                  options:NSStringDrawingUsesLineFragmentOrigin
                               attributes:atts
                                  context:context];
  return CGSizeMake((CGFloat) ceil(box.size.width), (CGFloat) ceil(box.size.height));
}

@end

@interface TKUITripSegmentsView ()

@property (nonatomic, assign) CGSize desiredSize;

@property (nonatomic, assign) BOOL didLayoutSubviews;
@property (nonatomic, nullable, strong) void (^onLayoutSubviews)(void);

@property (nonatomic, copy) NSArray<NSNumber *>* segmentXValues;
@property (nonatomic, assign) NSInteger segmentIndexToSelect;

@end

@implementation TKUITripSegmentsView

- (instancetype)initWithFrame:(CGRect)frame
{
  self = [super initWithFrame:frame];
  if (self) {
    [self didInit];
  }
  return self;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder
{
  self = [super initWithCoder:aDecoder];
  if (self) {
    [self didInit];
  }
  return self;
}

- (void)didInit {
  self.darkTextColor = UIColor.tkLabelPrimary;
  self.lightTextColor = UIColor.tkLabelSecondary;
  self.segmentIndexToSelect = -1;
}

- (CGSize)intrinsicContentSize
{
  if (CGSizeEqualToSize(CGSizeZero, self.desiredSize) && self.onLayoutSubviews != nil) {
    self.didLayoutSubviews = YES;
    self.onLayoutSubviews();
    self.onLayoutSubviews = nil;
  }
  return self.desiredSize;
}

- (void)layoutSubviews
{
  [super layoutSubviews];
  self.didLayoutSubviews = YES;
  if (self.onLayoutSubviews != nil) {
    self.onLayoutSubviews();
    self.onLayoutSubviews = nil;
  }
}

- (void)configureForSegments:(NSArray<id<TKTripSegmentDisplayable>> *)segments
              allowSubtitles:(BOOL)allowSubtitles
              allowInfoIcons:(BOOL)allowInfoIcons
{
  [self configureForSegments:segments allowTitles:allowSubtitles allowSubtitles:allowSubtitles allowInfoIcons:allowInfoIcons];
}

- (void)configureForSegments:(NSArray<id<TKTripSegmentDisplayable>> *)segments
                 allowTitles:(BOOL)allowTitles
              allowSubtitles:(BOOL)allowSubtitles
              allowInfoIcons:(BOOL)allowInfoIcons
{
  if (!self.didLayoutSubviews) {
    // When calling `configureForSegments` before `layoutSubviews` was called
    // the frame information of this view is most likely not yet what it'll
    // be before this view will be visible, so we'll delay configuring this view
    // until after `layoutSubviews` was called.
    __weak typeof(self) weakSelf = self;
    self.onLayoutSubviews = ^(){
      __strong typeof(weakSelf) strongSelf = weakSelf;
      [strongSelf configureForSegments:segments allowTitles:allowTitles allowSubtitles:allowSubtitles allowInfoIcons:allowInfoIcons];
    };
    return;
  }
  
  // clean up cells
  [self removeAllSubviews];
	
#define PADDING 12
  CGFloat padding = PADDING * (self.tiny ? 0.6f : 1.0f);
  CGFloat nextX = padding / 2;
  
  // We might not yet have a frame, if the Auto Layout engine wants to get the
  // intrinsic size of this view, before it's been added. In that case, we
  // let it grow as big as it wants to be.
  BOOL limitSize = !CGSizeEqualToSize(CGSizeZero, self.frame.size);
  CGFloat maxHeight = limitSize ? CGRectGetHeight(self.frame) : 44;
  
  NSMutableArray *segmentXValues = [NSMutableArray arrayWithCapacity:segments.count];
  NSInteger segmentIndexToSelect = (self.segmentIndexToSelect >=0 && self.segmentIndexToSelect < segments.count) ? self.segmentIndexToSelect : -1;
  
  // populate scroll view with segments
  int count = 0;
  
  for (id<TKTripSegmentDisplayable> segment in segments) {

    UIViewAutoresizing mask = 0;
    CGFloat alpha = (segmentIndexToSelect == -1 || segmentIndexToSelect == count) ? SEGMENT_ITEM_ALPHA_SELECTED : SEGMENT_ITEM_ALPHA_DESELECTED;
    
    // create image view
    UIImage *image = [segment tripSegmentModeImage];
		if (! image) {
      // this can happen if a trip was visible while TripKit got cleared
			continue;
		}
		
		// for the coloured strip
    UIColor *color = [segment tripSegmentModeColor];
		
		// the mode image
    UIImageView *modeImageView = [[UIImageView alloc] initWithImage:image];
    modeImageView.alpha = alpha;
    
    if (self.colorCodingTransitIcon) {
      // remember that color will be nil for non-PT modes. In these cases, since the
      // PT mode will be colored, we use the lighter grey to reduce the contrast.
      modeImageView.tintColor = color ?: self.lightTextColor;
    } else {
      modeImageView.tintColor = self.darkTextColor;
    }

    // place the mode image
    CGFloat imageWidth = image.size.width * (self.tiny ? 0.6f : 1.0f);
    CGFloat imageHeight = image.size.height * (self.tiny ? 0.6f : 1.0f);
    CGRect newFrame = modeImageView.frame;
    newFrame.origin.x = nextX;
    [segmentXValues addObject:@(nextX)];
    
    newFrame.origin.y = (maxHeight - imageHeight) / 2;
    newFrame.size.width = imageWidth;
    newFrame.size.height = imageHeight;
		modeImageView.autoresizingMask = mask;
    modeImageView.frame = newFrame;

    UIImageView *brandImageView;
    NSURL *modeImageURL = [segment tripSegmentModeImageURL];
    BOOL asTemplate = [segment tripSegmentModeImageIsTemplate];
    BOOL asBranding = [segment tripSegmentModeImageIsBranding];
    if (modeImageURL) {
      UIView *modeCircleBackground = nil;
      if (!asTemplate) {
        // remote images that aren't templates look weird on the background colour
        CGRect circleFrame = CGRectInset(newFrame, -1.0f, -1.0f);
        modeCircleBackground = [[UIView alloc] initWithFrame:circleFrame];
        modeCircleBackground.backgroundColor = [UIColor whiteColor];
        modeCircleBackground.layer.cornerRadius = CGRectGetWidth(circleFrame) / 2;
        modeCircleBackground.alpha = alpha;
        [self addSubview:modeCircleBackground];
      }
      
      if (asBranding) {
        // brand images are not overlaid over the mode icon, but appear next
        // to them
        CGRect brandFrame = newFrame;
        brandFrame.origin.x += brandFrame.size.width + 2;
        brandImageView = [[UIImageView alloc] initWithFrame:brandFrame];
        brandImageView.autoresizingMask = mask;
        brandImageView.alpha = alpha;
        [brandImageView setImageWithURL:modeImageURL];
        newFrame = brandFrame;
        
      } else {
        [modeImageView setImageWithURL:modeImageURL asTemplate:asTemplate placeholderImage:image completionHandler:^(BOOL succeed) {
          if (!succeed) {
            [modeCircleBackground removeFromSuperview];
          }
        }];
      }
      
      if (!asTemplate) {
        // add a little bit more spacing next to the circle background
        newFrame.origin.x += 2;
      }
    }

    [self addSubview:modeImageView];

    if (brandImageView) {
      [self addSubview:brandImageView];
    }

		if (allowInfoIcons && [segment tripSegmentModeInfoIconType] != TKInfoIconTypeNone) {
			UIImageView *infoIconImageView = [[UIImageView alloc] initWithImage:[TKInfoIcon imageForInfoIconType:[segment tripSegmentModeInfoIconType] usage:TKInfoIconUsageOverlay]];
      CGRect imageFrame = infoIconImageView.frame;
      imageFrame.origin.x = newFrame.origin.x;
      imageFrame.origin.y = newFrame.origin.y + CGRectGetHeight(newFrame) - CGRectGetHeight(imageFrame);
			infoIconImageView.frame = imageFrame;
			infoIconImageView.autoresizingMask = mask;
      infoIconImageView.alpha = modeImageView.alpha;
			[self addSubview:infoIconImageView];
		}
		
    __block CGFloat modeSideWidth = 0;
    // we put mode codes, colours and subtitles to the side
    // subtitle, we are allowed to
    
    CGFloat x = CGRectGetMaxX(newFrame);
    CGSize  modeTitleSize  = CGSizeZero;
    CGSize  modeSubtitleSize = CGSizeZero;
    
    UIFont *modeTitleFont = nil;
    UIFont *modeSubtitleFont = nil;
    
    NSString *modeTitle = allowTitles ? [segment tripSegmentModeTitle] : nil;
    if (modeTitle.length > 0) {
      modeTitleFont = [TKStyleManager systemFontWithSize:12];
      modeTitleSize = [modeTitle sizeWithFont:modeTitleFont
                           maximumWidth:CGFLOAT_MAX];
    } else if (color) {
      modeTitleSize = CGSizeMake(10, 10);
    }
    
    UIImageView *modeTitleAccessoryImageView = nil;
    if (allowTitles
        && self.allowWheelchairIcon
        && [segment tripSegmentIsWheelchairAccessible]) {
      modeTitleAccessoryImageView = [[UIImageView alloc] initAsWheelchairAccessoryImageWithTintColor:UIColor.tkLabelPrimary];
    }
    
    NSString *modeSubtitle = nil;
    NSMutableArray<UIImageView *> *modeSubtitleAccessoryImageViews = [NSMutableArray arrayWithCapacity:2];
    if (allowSubtitles) {
      // We prefer the time + real-time indicator as the subtitle, and fall back
      // to the subtitle
      NSDate *fixedTime = [segment tripSegmentFixedDepartureTime];
      if (fixedTime) {
        modeSubtitle = [TKStyleManager timeString:fixedTime
                                      forTimeZone:[segment tripSegmentTimeZone]];
      }

      if ([segment tripSegmentTimesAreRealTime]) {
        [modeSubtitleAccessoryImageViews addObject:[[UIImageView alloc] initAsRealTimeAccessoryImageAnimated:YES tintColor:UIColor.tkLabelSecondary]];
      }

      NSString *subtitleText = [segment tripSegmentModeSubtitle];
      if (! modeSubtitle && subtitleText.length > 0) {
        modeSubtitle = subtitleText;
      }
      
      if (modeSubtitle) {
        modeSubtitleFont = [TKStyleManager systemFontWithSize:10];
        modeSubtitleSize = [modeSubtitle sizeWithFont:modeSubtitleFont maximumWidth:CGFLOAT_MAX];
      }
      
      if (allowInfoIcons && [segment tripSegmentSubtitleIconType] != TKInfoIconTypeNone) {
        UIImageView *infoIconImageView = [[UIImageView alloc] initWithImage:[TKInfoIcon imageForInfoIconType:[segment tripSegmentSubtitleIconType] usage:TKInfoIconUsageOverlay]];
        [modeSubtitleAccessoryImageViews addObject:infoIconImageView];
      }
    }
    
    if (modeTitle.length > 0) {
      // add the bus number to the side
      CGFloat y = (maxHeight - modeSubtitleSize.height - modeTitleSize.height) / 2;
      if (modeSubtitleSize.height > 0 || modeSubtitleAccessoryImageViews.count > 0) {
        y += 2;
      }
      
      CGRect rect = CGRectMake(x + 2, y, modeTitleSize.width, modeTitleSize.height);
      TKUIStyledLabel *titleLabel = [[TKUIStyledLabel alloc] initWithFrame:rect];
      titleLabel.font = modeTitleFont;
      titleLabel.text = modeTitle;
      titleLabel.textColor = self.colorCodingTransitIcon ? self.lightTextColor : self.darkTextColor;
      titleLabel.alpha = modeImageView.alpha;
      [self addSubview:titleLabel];
      
      modeSideWidth = (CGFloat) fmax(modeSideWidth, modeTitleSize.width);
    } else if (color && allowSubtitles) {
      // add the color underneath
      CGFloat y = (maxHeight - modeSubtitleSize.height - modeTitleSize.height) / 2;
      UIView *stripe = [[UIView alloc] initWithFrame:CGRectMake(x, y, modeTitleSize.width, modeTitleSize.height)];
      stripe.layer.borderColor  = color.CGColor;
      stripe.layer.borderWidth  = modeTitleSize.width / 4;
      stripe.layer.cornerRadius = modeTitleSize.width / 2;
      stripe.alpha = modeImageView.alpha;
      [self addSubview:stripe];
      modeSideWidth = (CGFloat) fmax(modeSideWidth, modeTitleSize.width);
    }
    if (modeTitleAccessoryImageView) {
      CGSize imageSize = modeTitleAccessoryImageView.image.size;
      CGFloat viewWidth;
      CGFloat viewHeight;
      if (modeTitleSize.height > 0) {
        viewHeight = MIN(imageSize.height, modeTitleSize.height);
      } else {
        viewHeight = MIN(imageSize.height, 20);
      }
      viewWidth = viewHeight * (imageSize.width / imageSize.height);
      
      CGFloat y = (maxHeight - modeSubtitleSize.height - viewHeight) / 2;
      if (modeSubtitleSize.height > 0 || modeSubtitleAccessoryImageViews.count > 0) {
        y += 2;
      }
      
      modeTitleAccessoryImageView.frame = CGRectMake(x + modeTitleSize.width + 2, y, viewWidth, viewHeight);
      modeTitleAccessoryImageView.alpha = modeImageView.alpha;
      [self addSubview:modeTitleAccessoryImageView];
      modeSideWidth = (CGFloat) fmax(modeSideWidth, modeTitleSize.width + 2 + viewWidth);
    }
    
    
    if (modeSubtitle.length > 0) {
      // label goes under the mode code (if we have one)
      CGFloat y = (maxHeight - modeSubtitleSize.height - modeTitleSize.height) / 2 + modeTitleSize.height;
      
      CGRect rect = CGRectMake(x + 2, y, modeSubtitleSize.width, modeSubtitleSize.height);
      TKUIStyledLabel *subtitleLabel = [[TKUIStyledLabel alloc] initWithFrame:rect];
      subtitleLabel.font = modeSubtitleFont;
      subtitleLabel.text = modeSubtitle;
      subtitleLabel.textColor = UIColor.tkLabelSecondary;
      subtitleLabel.alpha = modeImageView.alpha;
      [self addSubview:subtitleLabel];
      
      modeSideWidth = (CGFloat) fmax(modeSideWidth, modeSubtitleSize.width);
    }
    if (modeSubtitleAccessoryImageViews.count > 0) {
      __block CGFloat subtitleWidth = modeSubtitleSize.width;
      [modeSubtitleAccessoryImageViews enumerateObjectsUsingBlock:^(UIImageView * _Nonnull imageView, NSUInteger idx, BOOL * _Nonnull stop) {
        CGSize imageSize = imageView.image.size;
        CGFloat viewWidth;
        CGFloat viewHeight;
        if (modeSubtitleSize.height > 0) {
          viewHeight = MIN(imageSize.height, modeSubtitleSize.height);
        } else {
          viewHeight = MIN(imageSize.height, 20);
        }
        viewWidth = viewHeight * (imageSize.width / imageSize.height);
        
        CGFloat y = (maxHeight - viewHeight - modeTitleSize.height) / 2 + modeTitleSize.height;
        
        imageView.frame = CGRectMake(x + subtitleWidth + 2, y, viewWidth, viewHeight);
        imageView.alpha = modeImageView.alpha;
        [self addSubview:imageView];
        modeSideWidth = (CGFloat) fmax(modeSideWidth, subtitleWidth + 2 + viewWidth);
        subtitleWidth += 2 + viewWidth;
      }];
    }
      
    nextX = CGRectGetMaxX(newFrame) + modeSideWidth + padding;
    count++;
    if (allowSubtitles) {
      self.desiredSize = CGSizeMake(nextX, maxHeight);
    }
    
    if (limitSize && nextX > CGRectGetWidth(self.frame)) {
      // can we shrink?
      if (allowSubtitles) {
        [self configureForSegments:segments
                       allowTitles:allowTitles
                    allowSubtitles:NO
                    allowInfoIcons:allowInfoIcons];
        return;
      
      } else if (allowTitles) {
        [self configureForSegments:segments
                       allowTitles:NO
                    allowSubtitles:NO
                    allowInfoIcons:allowInfoIcons];
        return;
      }
    }
  }
  
  self.segmentXValues = [segmentXValues copy];
}

- (void)selectSegmentAtIndex:(NSInteger)index
{
  self.segmentIndexToSelect = index;
  if (!self.segmentXValues) {
    return;
  }
  
  CGFloat minXToSelect = self.segmentXValues[index].doubleValue;
  CGFloat maxXToSelect = index + 1 < self.segmentXValues.count ? self.segmentXValues[index + 1].doubleValue : CGFLOAT_MAX;

  for (UIView *subview in self.subviews) {
    CGFloat midX = CGRectGetMidX(subview.frame);
    if (midX >= minXToSelect && midX < maxXToSelect) {
      subview.alpha = SEGMENT_ITEM_ALPHA_SELECTED;
    } else {
      subview.alpha = SEGMENT_ITEM_ALPHA_DESELECTED;
    }
  }
}

- (NSInteger)segmentIndexAtX:(CGFloat)x NS_SWIFT_NAME(segmentIndex(atX:))
{
  if (!self.segmentXValues) {
    return 0;
  }

  __block NSUInteger index = 0;
  [self.segmentXValues enumerateObjectsUsingBlock:^(NSNumber * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
    if (obj.doubleValue > x) {
      *stop = true;
    } else {
      index = idx;
    }
  }];
  return index;
}



#pragma mark - Private helpers

- (UIColor *)colorForBackground:(UIColor *)backgroundColor withDefault:(UIColor *)defaultColor
{
	if ([defaultColor isDark]) {
		if ([backgroundColor isDark]) {
			return [UIColor whiteColor];
		} else {
			return defaultColor;
		}
	} else {
		if ([backgroundColor isDark]) {
			return defaultColor;
		} else {
			return [UIColor blackColor];
		}
	}
}


@end

File: Sources/TripKitUI/views/TKUITripSegmentsView.swift
//
//  TKUITripSegmentsView.swift
//  TripKitUI-iOS
//
//  Created by Adrian Schönig on 31/1/21.
//  Copyright © 2021 SkedGo Pty Ltd. All rights reserved.
//

import UIKit

import TripKit

public class TKUITripSegmentsView : UIView {
  
  /// Whether the trip should be shown as cancelled, i.e., with a line through it
  /// - default: `false`
  public var isCanceled: Bool = false
  
  /// Whether to show wheelchair accessibility and inaccessibility icons
  /// - default: `false`
  public var allowWheelchairIcon: Bool = false
  
  /// This property determines if the transit icon in the view should be color coded.
  public var colorCodingTransitIcon: Bool = false
  
  /// This color is used for darker texts. In addition, this is also the color which
  /// will be used to tint the transport mode icons if `colorCodingTransitIcon` is
  /// set to `false`.
  /// - default: `UIColor.tkLabelPrimary`
  public var darkTextColor: UIColor = .tkLabelPrimary
  
  /// This color is used on lighter texts. In addition, this is also the color which
  /// will be used to tint non-PT modes if `colorCodingTransitIcon` is set to `YES`.
  /// - default: `UIColor.tkLabelSecondary`
  public var lightTextColor: UIColor = .tkLabelSecondary
  
  public override init(frame: CGRect) {
    super.init(frame: frame)
  }
  
  public required init?(coder: NSCoder) {
    super.init(coder: coder)
  }
  
  public func configure(_ segments: [TKUITripSegmentDisplayable], allowSubtitles: Bool = true, allowInfoIcons: Bool = true) {
    configure(segments, allowTitles: allowSubtitles, allowSubtitles: allowSubtitles, allowInfoIcons: allowInfoIcons)
  }
  
  public func selectSegment(atIndex index: Int) {
    segmentIndexToSelect = index
    guard index >= 0, index < segmentLeftyValues.count else { return }
    
    let minX = segmentLeftyValues[index]
    let maxX = index + 1 < segmentLeftyValues.count ? segmentLeftyValues[index + 1] : .greatestFiniteMagnitude
    let isRightToLeft = effectiveUserInterfaceLayoutDirection == .rightToLeft
    for view in subviews {
      let midX = isRightToLeft ? bounds.width - view.frame.midX : view.frame.midX
      view.alpha = midX >= minX && midX < maxX ? Self.alphaSelected : Self.alphaDeselected
    }
  }
  
  public func segmentIndex(atX x: CGFloat) -> Int {
    let target = effectiveUserInterfaceLayoutDirection == .leftToRight ? x : bounds.width - x
    return segmentLeftyValues.lastIndex { $0 <= target } ?? 0
  }
  
  public override var intrinsicContentSize: CGSize {
    if desiredSize == .zero, let onLayout = self.onLayout {
      didLayoutSubviews = true
      onLayout()
      self.onLayout = nil
    }
    return desiredSize
  }
  
  public override func layoutSubviews() {
    super.layoutSubviews()
    didLayoutSubviews = true
    if let onLayout = self.onLayout {
      onLayout()
      self.onLayout = nil
    }
  }
  
  // MARK: - Constants
  
  private static let alphaSelected: CGFloat = 1
  private static let alphaDeselected: CGFloat = 0.25
  private static let padding: CGFloat = 12
  
  // MARK: - Internals
  
  private var desiredSize: CGSize = .zero
  private var didLayoutSubviews: Bool = false
  private var onLayout: (() -> Void)? = nil
  
  /// These are from the left, i.e., same for left-to-right and right-to-left
  private var segmentLeftyValues: [CGFloat] = []
  
  private var segmentIndexToSelect: Int? = nil
  
  private func configure(_ segments: [TKUITripSegmentDisplayable], allowTitles: Bool, allowSubtitles: Bool, allowInfoIcons: Bool) {
    if !didLayoutSubviews {
      // When calling `configureForSegments` before `layoutSubviews` was called
      // the frame information of this view is most likely not yet what it'll
      // be before this view will be visible, so we'll delay configuring this view
      // until after `layoutSubviews` was called.
      self.onLayout = { [weak self] in
        self?.configure(segments, allowTitles: allowTitles, allowSubtitles: allowSubtitles, allowInfoIcons: allowInfoIcons)
      }
      return
    }
    
    subviews.forEach { $0.removeFromSuperview() }

    var accessibileElements: [UIAccessibilityElement] = []
    
    var nextX = Self.padding / 2
    
    // We might not yet have a frame, if the Auto Layout engine wants to get the
    // intrinsic size of this view, before it's been added. In that case, we
    // let it grow as big as it wants to be.
    let limitSize = frame.size != .zero
    let maxHeight = limitSize ? frame.height : 44
    
    var newSegmentXValues: [CGFloat] = []
    let segmentIndexToSelect = self.segmentIndexToSelect.flatMap { $0 < segments.count ? $0 : nil }
    
    let isRightToLeft = effectiveUserInterfaceLayoutDirection == .rightToLeft
    contentMode = isRightToLeft ? .right : .left
    autoresizingMask = isRightToLeft ? [.flexibleLeftMargin, .flexibleBottomMargin] : [.flexibleRightMargin, .flexibleBottomMargin]
    
    var count = 0
    for segment in segments {
      let mask: UIView.AutoresizingMask = autoresizingMask
      let isSelected = (segmentIndexToSelect ?? count) == count
      let alpha = isSelected ? Self.alphaSelected : Self.alphaDeselected
      
      guard let modeImage = segment.tripSegmentModeImage else {
        continue // this can happen if a trip was visible while TripKit got cleared
      }
      
      // 1. The mode and brand images, maybe with circled behind them
      
      let color = segment.tripSegmentModeColor
      let modeImageView = UIImageView(image: modeImage)
      modeImageView.autoresizingMask = mask
      modeImageView.alpha = alpha
      
      // remember that color will be nil for non-PT modes. In these cases, since the
      // PT mode will be colored, we use the lighter grey to reduce the contrast.
      modeImageView.tintColor = colorCodingTransitIcon ? (color ?? lightTextColor) : darkTextColor
      
      newSegmentXValues.append(nextX)
      modeImageView.frame.origin.x = nextX
      modeImageView.frame.origin.y = (maxHeight - modeImage.size.height) / 2
      
      var newFrame = modeImageView.frame
      
      var brandImageView: UIImageView? = nil
      if let modeImageURL = segment.tripSegmentModeImageURL {
        let asTemplate = segment.tripSegmentModeImageIsTemplate
        
        @discardableResult
        func addCircle(frame: CGRect) -> UIView {
          let circleFrame = frame.insetBy(dx: -1, dy: -1)
          let modeCircleBackground = UIView(frame: circleFrame)
          modeCircleBackground.autoresizingMask = mask
          modeCircleBackground.backgroundColor = .white
          modeCircleBackground.layer.cornerRadius = circleFrame.width / 2
          modeCircleBackground.alpha = alpha
          addSubview(modeCircleBackground)
          return modeCircleBackground
        }
        
        if segment.tripSegmentModeImageIsBranding {
          var brandFrame = newFrame
          brandFrame.origin.x += brandFrame.width + 4
          
          // Always add a circle first as these look weird on background color
          addCircle(frame: brandFrame)
          
          // brand images are not overlaid over the mode icon, but appear next
          // to them
          let brandImage = UIImageView(frame: brandFrame)
          brandImage.autoresizingMask = mask
          brandImage.alpha = alpha
          brandImage.setImage(with: modeImageURL, asTemplate: false)
          brandImageView = brandImage
          newFrame = brandFrame

        } else {
          // remote images that aren't templates look weird on the background colour
          let modeCircleBackground = asTemplate ? nil : addCircle(frame: newFrame)
          modeImageView.setImage(with: modeImageURL, asTemplate: asTemplate, placeholder: modeImage) { succeeded in
            guard succeeded else { return }
            modeCircleBackground?.removeFromSuperview()
          }
        }
        if !asTemplate {
          // add a little bit more spacing next to the circle background
          newFrame.origin.x += 2
        }
      }
      
      addSubview(modeImageView)
      if let brandImageView = brandImageView {
        addSubview(brandImageView)
      }
      
      // 2. Optional info icon
      if allowInfoIcons, let image = TKInfoIcon.image(for: segment.tripSegmentModeInfoIconType, usage: .overlay) {
        let infoIconImageView = UIImageView(image: image)
        infoIconImageView.autoresizingMask = mask
        infoIconImageView.frame.origin.x = newFrame.minX
        infoIconImageView.frame.origin.y = newFrame.maxY - image.size.height
        infoIconImageView.alpha = modeImageView.alpha
        addSubview(infoIconImageView)
      }
      
      // we put mode codes, colours and subtitles to the side
      // subtitle, we are allowed to
      var modeSideWith: CGFloat = 0
      let x = newFrame.maxX
      var modeSubtitleSize = CGSize.zero
      let modeTitleFont = TKStyleManager.customFont(forTextStyle: .caption1)
      let modeSubtitleFont = TKStyleManager.customFont(forTextStyle: .caption2)
      
      let modeTitle = allowTitles ? segment.tripSegmentModeTitle.flatMap { $0.isEmpty ? nil : $0 } : nil
      let modeTitleSize = modeTitle?.size(font: modeTitleFont)
        ?? color.map { _ in .init(width: 10, height: 10) }
        ?? .zero
      
      let modeTitleAccessoryImageView = allowTitles && allowWheelchairIcon ? TKUISemaphoreView.accessibilityImageView(for: segment) : nil
 
      var modeSubtitle: String? = nil
      var modeSubtitleAccessoryImageViews: [UIImageView] = []
      if allowSubtitles {
        // We prefer the time + real-time indicator as the subtitle, and fall back
        // to the subtitle
        if let fixedTime = segment.tripSegmentFixedDepartureTime {
          modeSubtitle = TKStyleManager.timeString(fixedTime, for: segment.tripSegmentTimeZone)
        }
        if segment.tripSegmentTimesAreRealTime {
          modeSubtitleAccessoryImageViews.append(UIImageView(asRealTimeAccessoryImageAnimated: true, tintColor: lightTextColor))
        }
        if modeSubtitle == nil, let subtitle = segment.tripSegmentModeSubtitle, !subtitle.isEmpty {
          modeSubtitle = subtitle
        }
        if let subtitle = modeSubtitle {
          modeSubtitleSize = subtitle.size(font: modeSubtitleFont)
        }
        if allowInfoIcons, let subtitleIcon = TKInfoIcon.image(for: segment.tripSegmentSubtitleIconType, usage: .overlay) {
          modeSubtitleAccessoryImageViews.append(UIImageView(image: subtitleIcon))
        }
      }
      
      if let modeTitle = modeTitle {
        var y = (maxHeight - modeSubtitleSize.height - modeTitleSize.height) / 2
        if modeSubtitleSize.height > 0 || !modeSubtitleAccessoryImageViews.isEmpty {
          y += 2
        }
        let label = TKUIStyledLabel(frame: .init(origin: .init(x: x + 2, y: y), size: modeTitleSize))
        label.autoresizingMask = mask
        label.font = modeTitleFont
        label.text = modeTitle
        label.textColor = colorCodingTransitIcon ? lightTextColor : darkTextColor
        label.alpha = modeImageView.alpha
        addSubview(label)
        modeSideWith = max(modeSideWith, modeTitleSize.width)
      
      } else if allowSubtitles, let color = color {
        let y = (maxHeight - modeSubtitleSize.height - modeTitleSize.height) / 2
        let stripe = UIView(frame: .init(origin: .init(x: x, y: y), size: modeTitleSize))
        stripe.autoresizingMask = mask
        stripe.layer.borderColor = color.cgColor
        stripe.layer.borderWidth = modeTitleSize.width / 4
        stripe.layer.cornerRadius = modeTitleSize.width / 2
        stripe.alpha = modeImageView.alpha
        addSubview(stripe)
        modeSideWith = max(modeSideWith, modeTitleSize.width)
      }
      
      if let accessoryImageView = modeTitleAccessoryImageView, let accessoryImage = accessoryImageView.image {
        let viewHeight = modeTitleSize.height > 0
          ? min(accessoryImage.size.height, modeTitleSize.height)
          : min(accessoryImage.size.height, 20)
        let viewWidth = viewHeight * accessoryImage.size.width / accessoryImage.size.height
        
        var y = (maxHeight - modeSubtitleSize.height - viewHeight) / 2
        if modeSubtitleSize.height > 0 || !modeSubtitleAccessoryImageViews.isEmpty {
          y += 2
        }
        accessoryImageView.frame = .init(x: x + modeTitleSize.width + 2, y: y, width: viewWidth, height: viewHeight)
        accessoryImageView.alpha = modeImageView.alpha
        addSubview(accessoryImageView)
        modeSideWith = max(modeSideWith, modeTitleSize.width + 2 + viewWidth)
      }
      
      if let subtitle = modeSubtitle, !subtitle.isEmpty {
        // label goes under the mode code (if we have one)
        let y = (maxHeight - modeSubtitleSize.height - modeTitleSize.height) / 2 + modeTitleSize.height
        let label = TKUIStyledLabel(frame: .init(origin: .init(x: x + 2, y: y), size: modeSubtitleSize))
        label.autoresizingMask = mask
        label.font = modeSubtitleFont
        label.text = subtitle
        label.textColor = lightTextColor
        label.alpha = modeImageView.alpha
        addSubview(label)
        modeSideWith = max(modeSideWith, modeSubtitleSize.width)
      }
      
      var subtitleWidth = modeSubtitleSize.width
      for imageView in modeSubtitleAccessoryImageViews {
        guard let image = imageView.image else { assertionFailure(); continue }
        imageView.autoresizingMask = mask
        let viewHeight = modeSubtitleSize.height > 0
          ? min(image.size.height, modeSubtitleSize.height)
          : min(image.size.height, 20)
        let viewWidth = viewHeight * image.size.width / image.size.height
        let y = (maxHeight - viewHeight - modeTitleSize.height) / 2 + modeTitleSize.height
        imageView.frame = .init(x: x + subtitleWidth + 2, y: y, width: viewWidth, height: viewHeight)
        imageView.alpha = modeImageView.alpha
        addSubview(imageView)
        modeSideWith = max(modeSideWith, subtitleWidth + 2 + viewWidth)
        subtitleWidth += 2 + viewWidth
      }
      
      newFrame.size.width += modeSideWith
      
      let accessibleElement = UIAccessibilityElement(accessibilityContainer: self)
      accessibleElement.accessibilityLabel = segment.tripSegmentAccessibilityLabel
      accessibleElement.accessibilityFrameInContainerSpace = newFrame
      if segmentIndexToSelect != nil {
        accessibleElement.accessibilityTraits = isSelected ? [.button, .selected] : [.button]
      }
      accessibileElements.append(accessibleElement)
      
      nextX = newFrame.maxX + Self.padding
      count += 1
      if allowSubtitles {
        desiredSize = .init(width: nextX, height: maxHeight)
      }
      if limitSize, nextX > frame.width {
        // try to shrink
        if allowSubtitles {
          configure(segments, allowTitles: allowTitles, allowSubtitles: false, allowInfoIcons: allowInfoIcons)
          return
        } else if allowTitles {
          configure(segments, allowTitles: false, allowSubtitles: false, allowInfoIcons: allowInfoIcons)
          return
        }
      }
      
    }
    
    if isCanceled {
      let lineHeight: CGFloat = 1
      let strikethrough = UIView(frame: .init(x: 0, y: (maxHeight - lineHeight) / 2, width: nextX, height: lineHeight))
      strikethrough.backgroundColor = darkTextColor
      addSubview(strikethrough)
    }
        
    if isRightToLeft {
      for view in subviews {
        view.frame.origin.x = frame.width - view.frame.maxX
      }
    }

    self.segmentLeftyValues = newSegmentXValues
    
    if segmentIndexToSelect != nil {
      self.accessibilityElements = accessibileElements
      self.isAccessibilityElement = false
    }
  }
  
}

extension String {
  fileprivate func size(font: UIFont, maximumWidth: CGFloat = .greatestFiniteMagnitude) -> CGSize {
    let context = NSStringDrawingContext()
    context.minimumScaleFactor = 1
    let box = (self as NSString).boundingRect(
      with: .init(width: maximumWidth, height: 0),
      options: .usesLineFragmentOrigin,
      attributes: [.font: font],
      context: context
    )
    return .init(
      width: box.width.rounded(.up),
      height: box.height.rounded(.up)
    )
  }
}
